File: Benchmark-Application/benchmarks/ThreadContextSwitch.py
import threading
import time
from datetime import datetime

iterations = 100
NUMBER_OF_THREADS = 10000
counter = 0
cout_mutex = threading.Lock()


def thread_function(id, iterations):
    global counter
    for _ in range(iterations):
        with cout_mutex:
            counter += 1
        time.sleep(0)


def exec_program():
    # Record the start time
    start_time = time.time()

    # Declare an array of thread objects
    threads = []

    # Start the threads
    for i in range(NUMBER_OF_THREADS):
        t = threading.Thread(target=thread_function, args=(i, iterations))
        threads.append(t)
        t.start()

    # Join the threads
    for t in threads:
        t.join()

    # Calculate the elapsed time in seconds
    end_time = time.time()
    elapsed_time = end_time - start_time

    print(f"{elapsed_time:.2e}")


if __name__ == "__main__":
    exec_program()

File: Benchmark-Application/benchmarks/ThreadContextSwitch.cpp
#include <iostream>
#include <thread>
#include <mutex>
#include <chrono>

using namespace std;
using namespace chrono;

const int iterations = 100;
const int NUMBER_OF_THREADS = 10000;
mutex coutMutex;

long counter = 0;

void threadFunction(int id, int iterations)
{
    for (int i = 0; i < iterations; ++i)
    {
        lock_guard<mutex> lock(coutMutex);
        counter++;
        this_thread::yield();
    }
}

void exec()
{
    // Record the start time
    high_resolution_clock::time_point start_time = high_resolution_clock::now();

    // Declare an array of thread objects
    thread threads[NUMBER_OF_THREADS];

    // Start the threads
    for (int i = 0; i < NUMBER_OF_THREADS; i++)
    {
        threads[i] = thread(threadFunction, i, iterations);
    }

    // Join the threads
    for (int i = 0; i < NUMBER_OF_THREADS; i++)
    {
        threads[i].join();
    }

    // Calculate the elapsed time in seconds
    high_resolution_clock::time_point end_time = high_resolution_clock::now();
    duration<double> elapsed_time = duration_cast<duration<double>>(end_time - start_time);

    cout << scientific << elapsed_time.count();
}

int main()
{
    exec();

    return 0;
}
